pyxdf/pyxdf.py

```diff
@@ -170,10 +170,10 @@ class StreamData:
         """Temporary per-stream data."""
         def __init__(self, xml):
             """Init a new StreamData object from a stream header."""
-            fmt2char = {'int8': 'b', 'int16': 'h', 'int32': 'i', 'int64': 'q',
-                        'float32': 'f', 'double64': 'd'}
-            fmt2nbytes = {'int8': 1, 'int16': 2, 'int32': 4, 'int64': 8,
-                          'float32': 4, 'double64': 8}
+            fmts = ['double64', 'float32', None, 'int32', 'int16', 'int8', 'int64']
```

Comment:
A dict seems like a more natural container here.

`fmts = {'double64': {'char': 'd', 'nbytes': 8, 'dtype': np.float64}, 'float32': 'char': 'f', 'nbytes': 4, 'dtype': np.float32}}`

With better white spacing, the above is easier to understand and use in the code below. It's easier to maintain in that you don't have to pay attention to ordering in 4 different lists, and, while typing keys is admittedly a downside, doing it incorrectly should generate an error immediately, whereas incorrect ordering will be a silent bug.

Reply:
Agreed!

Reply:
Ok, guess I'm overruled :-)

In order to make things simple and foolproof, I have replaced out the call to `struct.unpack` with `numpy.frombuffer` so we don't need the char argument. I'd have taken `numpy.dtype.char`, but struct and numpy have different opinions what character int64 should have.

fmts is now an OrderedDict so we get the benefits of both worlds.

Reply:
Please don't use an OrderedDict if you don't really need it. `self.fmtidx` is never needed, because you should index with the dict's keys.

Reply:
In the chunk7 proposal I have one field for the LSL type index (which is an uint8), so in the future it is needed.

Reply:
Why is this necessary? The format is already defined in the stream header, and having another place where this is stored is not ideal. In addition, I also don't like the idea that the index of a list is meaningful.

Reply:
It's one additional byte in the chunk header that has to be saved and processed. On the other hand, 

- each chunk is self-contained (all necessary informations like dimensions and data format to read the chunk data are in the chunk header),
- it allows additional level of error-checking by comparing the chunk header with the information in the stream header, either if the XDF writer changed the format mid-stream (unlikely) or the reader was left in the wrong position and continues to interpret stream data as chunk information.

The full type name would be better / more informative but it would either require a length and then the type name or a fixed length type name that might not be long enough if other types are added in the future.

Reply:
XDF always consists of multiple chunks, so a chunk need not be self-contained. It's not about the additional byte in the chunk7 header, but we already have a format definition, and having it in two places will create maintenance problems at some point. Therefore, I'd remove this information from the chunk7 header, which will make this code unnecessary too.

Reply:
I still think it's a good idea, but unrelated to this PR so I have removed it.

Reply:
Thanks, we can continue discussing this in the chunk7 proposal.